core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.http.dynamic;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * A bindable path. It holds three things:
+ * <ul>
+ *   <li> A regex-compiled form of the path. It is used for matching and binding.
+ *   <li> A skeleton of the path. It is used for duplication detecting.
+ *   <li> A set of variables declared in the path.
+ * </ul>
+ *
+ * @see DynamicHttpService
+ * @see Path
+ */
+final class DynamicPath {
+
+    private static final Pattern VALID_PATTERN = Pattern.compile("(/[^/]+)+");
+
+    /**
+     * Create a {@link DynamicPath} instance from given str.
```

Comment:
I also fixed the constructor of `DynamicPath`, along with the static function you mentioned.